### PR TITLE
Require `SizeConst` for `UnmanagedType.ByValArray`

### LINKED ISSUE #68988: Roslyn requires the SizeConst metadata on a MarshalAsAttribute for UnmanagedType.ByValTStr but not UnmanagedType.ByValArray
**Version Used**: 4.7.0

**Steps to Reproduce**:

1.  Add the `[MarshalAs(UnmanagedType.ByValTStr)]` attribute to a field.
2.  Observe that a CS7046 error is reported.
3.  Change the `UnmanagedType` to `UnmanagedType.ByValArray`.
4. Observe that the error goes away even though all systems have the same requirements for both.

A minimal repro, with source-code provided, is ideal.  Using [sharplab](https://sharplab.io/) is preferred for compiler/language issues whenever possible.

[SharpLab Link](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgBgATIIwDoBKArgHbBgC2ApjgJIkUBOA9gA4DKDAbmAMYUDOAbgCwAKFHIAzOj7B6BbsHStRAb1HoN6ANoBZAIb0+ACz0QAgnwAUAVSJk9RPQHMKMACoBPZlQBCHgGqmbqyyAJQAuuqakphYGIJRGona+oYm5la29o4u7l6+AaZm9PR6HhHJMWAkWuHoesIimuiiAL5AA=)

**Diagnostic Id**: CS7046

If this is a report about a bug in an analyzer, please include the diagnostic if possible (e.g. `"IDE0030"`).

**Expected Behavior**:

CS7046 is reported for both `ByValTStr` and `ByValArray`, as all runtimes (.NET Framework, CoreCLR, etc) require the SizeConst metadata and throw an exception if it is not present for both cases.

**Actual Behavior**:

CS7046 is only reported for `ByValTStr`.

Fixing this issue would obsolete https://github.com/dotnet/runtime/issues/36134 (as the compiler already does half the analyzer, and this issue corresponds to the remaining work).

### PR DESCRIPTION
**Pull request by jkoritzinsky**
`MarshalAsAttribute.SizeConst` is required by all runtimes for `UnmanagedType.ByValArray`, so warn when it is not provided.

As mentioned below, we cannot make it an error due to back-compat.

Fixes #68988
Fixes dotnet/runtime#36134

### CODE REVIEW

**Review by jaredpar — DISMISSED** [review]
This cannot be an error

File: src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb [hunk 1]
```diff
@@ -622,6 +622,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
+        Public Overrides ReadOnly Property WRN_ByValArraySizeConstRequired As Integer
+            Get
+                Return ERRID.Unknown
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb:627]: To my knowledge, VB doesn't have warning waves, so I can't do a warning diagnostic here. I wasn't sure what diagnostic code to use here for "no diagnostic". I can make it nullable instead.

**Comment by jkoritzinsky** [comment]
@jaredpar @AlekseyTs can I get another review pass?

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 2]
```diff
@@ -7526,6 +7526,12 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="WRN_AddressOfInAsync_Title" xml:space="preserve">
     <value>The '&amp;' operator should not be used on parameters or local variables in async methods.</value>
   </data>
+  <data name="WRN_ByValArraySizeConstRequired" xml:space="preserve">
+    <value>Attribute parameter 'SizeConst' must be specified.</value>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:7530]: >parameter [](http://example.com/codeflow?start=21&length=9)

SizeConst is a field, not a parameter. Right?

**Comment by AlekseyTs** [comment]
>         public void NativeTypeFixedArray()

Instead of modifying attributes in this test, we should assert the warnings. Otherwise we are not verifying metadata emitted for the affected scenario.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests_MarshalAs.cs:565 in 60d535c. [](commit_id = 60d535c7e1f808cfedd541289bf1a219cc742854, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         public void NativeTypeFixedArray()

Also should test that the warning is suppressed if ```CompilationOptions.WarningLever < 8``` is used.

---
In reply to: [1644180537](https://github.com/dotnet/roslyn/pull/68992#issuecomment-1644180537) [](http://example.com/codeflow?ancestors=1644180537)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests_MarshalAs.cs:565 in 60d535c. [](commit_id = 60d535c7e1f808cfedd541289bf1a219cc742854, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         public void NativeTypeFixedArray()

We should check if any diagnostic is reported for similar VB test.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests_MarshalAs.cs:565 in 60d535c. [](commit_id = 60d535c7e1f808cfedd541289bf1a219cc742854, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 10)

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 2 above)
Comment (jkoritzinsky) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:7530]: It's a field, but I wanted to have this message match the message for when `SizeConst` is not specified in the `ByValTStr` case (which uses the message in `ERR_AttributeParameterRequired1`).

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:7530]: > It's a field, but I wanted to have this message match the message for when SizeConst is not specified in the ByValTStr case (which uses the message in ERR_AttributeParameterRequired1).

Fair enough.

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs [hunk 3]
```diff
@@ -209,6 +209,7 @@ internal static int GetWarningLevel(ErrorCode code)
             switch (code)
             {
                 case ErrorCode.WRN_AddressOfInAsync:
+                case ErrorCode.WRN_ByValArraySizeConstRequired:
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:212]: Yes, it is tested here: https://github.com/dotnet/roslyn/pull/68992/files#diff-4604a322566a4681e081c334cb9516691a7308ac601887932d0d76d34b34b62dR449

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 12)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Pull request merged**